hummingbot/connector/exchange/paper_trade/paper_trade_exchange.pyx

```diff
@@ -438,40 +434,73 @@ cdef class PaperTradeExchange(ExchangeBase):
                                   order_id)))
         return order_id
 
-    cdef c_execute_buy(self, str order_id, str trading_pair, object amount):
+    cdef c_execute_buy(self, str order_id, str trading_pair_str, object amount):
         cdef:
-            str quote_asset = self._trading_pairs[trading_pair].quote_asset
-            str base_asset = self._trading_pairs[trading_pair].base_asset
+            str quote_asset = self._trading_pairs[trading_pair_str].quote_asset
+            str base_asset = self._trading_pairs[trading_pair_str].base_asset
             object quote_balance = self.c_get_balance(quote_asset)
             object base_balance = self.c_get_balance(base_asset)
 
-        config = self._config
-        order_book = self.order_books[trading_pair]
+        order_book = self.order_books[trading_pair_str]
+
         buy_entries = order_book.simulate_buy(amount)
-        # Calculate the quote currency needed, including fees.
-        total_quote_needed = Decimal(sum(row.price * row.amount for row in buy_entries))
 
-        if total_quote_needed > quote_balance:
+        # Get the weighted average price of the trade
+        avg_price = Decimal(0)
+        for entry in buy_entries:
+            avg_price += Decimal(entry.price) * Decimal(entry.amount)
+        avg_price = avg_price / amount
+
+        order_candidate = OrderCandidate(
+            trading_pair=trading_pair_str,
+            # Market orders are not maker orders
+            is_maker=False,
+            order_type=OrderType.MARKET,
+            order_side=TradeType.BUY,
+            amount=amount,
+            price=avg_price
+        )
+
+        adjusted_candidate_order = self._budget_checker.adjust_candidate(order_candidate, all_or_none=True)
+
+        # Base currency acquired, including fees.
+        sold_amount = adjusted_candidate_order.order_collateral.amount
```

Comment:
The name `sold_amount` is confusing since we are processing a buy.

Reply:
sold_amount and acquited_amount are used in both execute_buy() and execute_sell(), each time one asset has to be sold for the other to be bought and vice versa.